pkg/controller/admissionchecks/multikueue/controller.go

```diff
@@ -0,0 +1,380 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package multikueue
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	apimeta "k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	kueuealpha "sigs.k8s.io/kueue/apis/kueue/v1alpha1"
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1beta1"
+	"sigs.k8s.io/kueue/pkg/util/admissioncheck"
+)
+
+const (
+	ControllerName = "kueue.x-k8s.io/multikueue"
+)
+
+type multiKueueStoreHelper = admissioncheck.ConfigHelper[*kueuealpha.MultiKueueConfig, kueuealpha.MultiKueueConfig]
+
+func newMultiKueueStoreHelper(c client.Client) (*multiKueueStoreHelper, error) {
+	return admissioncheck.NewConfigHelper[*kueuealpha.MultiKueueConfig](c)
+}
+
+// AcReconciler implements the reconciler for all the admission checks controlled by multikueue.
+// Its main tasks being to:
+// - Maintain the list of remote controllers associated to each admission checks.
+// - Maintain the active state of the admission checks.
+type AcReconciler struct {
+	client client.Client
+	helper *multiKueueStoreHelper
+
+	lock sync.RWMutex
+	// The list of remote controllers, indexed by the admission checks name.
+	controllers map[string]*remoteController
+	wlUpdateCh  chan event.GenericEvent
+
+	rootContext context.Context
+
+	// For testing only.
+	updateConfigOverride func(ctx context.Context, rc *remoteController, kubeconfigs map[string][]byte) error
```

Comment:
Can you explain why this is needed for testing? Is there another workaround feasible, could we use fake client?

Reply:
done